### PR TITLE
Account for short-hand field syntax when suggesting borrow

### LINKED ISSUE #52965: Incorrect dereferencing suggestion when using init shorthand
[The following suggestion](https://play.rust-lang.org/?version=stable&mode=debug&edition=2015&gist=23c6eff2922fb34553a465fee5917006) is incorrect:

```
error[E0308]: mismatched types
   -->
    |
793 |         Condition::Function { function, description, expressions: exprs? }
    |                               ^^^^^^^^
    |                               |
    |                               expected enum `policy::MatchFunction`, found reference
    |                               help: consider dereferencing the borrow: `*function`
    |
    = note: expected type `policy::MatchFunction`
               found type `&policy::MatchFunction`
```

As the code is using [the init shorthand syntax](https://doc.rust-lang.org/book/second-edition/ch05-01-defining-structs.html#using-the-field-init-shorthand-when-variables-and-fields-have-the-same-name) to create the `Function` variant, the suggestion should instead be:

```
error[E0308]: mismatched types
   -->
    |
793 |         Condition::Function { function, description, expressions: exprs? }
    |                               ^^^^^^^^
    |                               |
    |                               expected enum `policy::MatchFunction`, found reference
    |                               help: consider dereferencing the borrow: `function: *function`
    |
    = note: expected type `policy::MatchFunction`
               found type `&policy::MatchFunction`
```

### PR DESCRIPTION
**Pull request by estebank**
Fix #52965.

### CODE REVIEW

File: src/librustc_typeck/check/demand.rs [hunk 1]
```diff
@@ -277,6 +277,24 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
             return None;
         }
 
+        let parent_id = self.tcx.hir().get_parent_node_by_hir_id(expr.hir_id);
```
Comment (Centril) [inline src/librustc_typeck/check/demand.rs]: Extract to function? (`fn is_struct_pat_shorthand_field(...) -> bool`)

File: src/librustc_typeck/check/demand.rs [hunk 2]
```diff
@@ -341,14 +359,22 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
                         if let Some(sugg) = self.can_use_as_ref(expr) {
                             return Some(sugg);
                         }
-                        return Some(match mutability {
-                            hir::Mutability::MutMutable => {
-                                (sp, "consider mutably borrowing here", format!("&mut {}",
-                                                                                sugg_expr))
+                        return Some(match (mutability, is_struct_pat_shorthand_field) {
```
Comment (Centril) [inline src/librustc_typeck/check/demand.rs]: I'd personally extract this as well... :) `fn field_shorthand_suggestion()` or some such

**Review by davidtwco — APPROVED** [review]
LGTM. Once @Centril's comments are resolved, r=me.

File: src/librustc_typeck/check/demand.rs [hunk 3]
```diff
@@ -249,6 +249,26 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
         None
     }
 
+    fn is_hir_id_from_struct_pattern_shorthand_field(&self, hir_id: hir::HirId) -> bool {
+        let parent_id = self.tcx.hir().get_parent_node_by_hir_id(expr.hir_id);
+        let mut is_struct_pat_shorthand_field = false;
```
Comment (Centril) [inline src/librustc_typeck/check/demand.rs]:
```suggestion
```

File: src/librustc_typeck/check/demand.rs [hunk 4]
```diff
@@ -249,6 +249,26 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
...
+        if let Some(parent) = self.tcx.hir().find_by_hir_id(parent_id) {
+            // Account for fields
+            if let Node::Expr(hir::Expr {
+                node: hir::ExprKind::Struct(_, fields, ..), ..
+            }) = parent {
+                if let Ok(src) = cm.span_to_snippet(sp) {
+                    for field in fields {
+                        if field.ident.as_str() == src.as_str() && field.is_shorthand {
+                            is_struct_pat_shorthand_field = true;
```
Comment (Centril) [inline src/librustc_typeck/check/demand.rs]:
```suggestion
                            return true;
```

File: src/librustc_typeck/check/demand.rs [hunk 5]
```diff
@@ -249,6 +249,26 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
...
+                            break;
```
Comment (Centril) [inline src/librustc_typeck/check/demand.rs]:
```suggestion
```

File: src/librustc_typeck/check/demand.rs [hunk 6]
```diff
@@ -249,6 +249,26 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
...
+                        }
+                    }
+                }
+            }
+        };
```
Comment (Centril) [inline src/librustc_typeck/check/demand.rs]:
```suggestion
        }

        false
```

**Pull request merged**